src/core/thread/lowpan.cpp

```diff
@@ -408,11 +410,54 @@ int Lowpan::CompressExtensionHeader(Message &aMessage, uint8_t *aBuf, uint8_t &a
     cur++;
 
     len = (extHeader.GetLength() + 1) * 8 - sizeof(extHeader);
+
+    // RFC 6282 says: "IPv6 Hop-by-Hop and Destination Options Headers may use a trailing
+    // Pad1 or PadN to achieve 8-octet alignment. When there is a single trailing Pad1 or PadN
+    // option of 7 octets or less and the containing header is a multiple of 8 octets, the trailing
+    // Pad1 or PadN option MAY be elided by the compressor."
+    if (aNextHeader == Ip6::kProtoHopOpts || aNextHeader == Ip6::kProtoDstOpts)
+    {
+        // Read last 7-bytes from extension header.
+        aMessage.Read(aMessage.GetOffset() + (8 * extHeader.GetLength()) - 1, 7, padBytes);
+
+        // Count number of "0" by analyzing the last 5-octets starting from the last one.
+        for (zeroCount = 0; zeroCount < 6; zeroCount++)
+        {
+            if (padBytes[6 - zeroCount] != Ip6::OptionPadN::kData)
+            {
+                break;
+            }
+        }
+
+        if (zeroCount != 0)
+        {
+            if ((padBytes[6 - zeroCount] == zeroCount) &&
+                (padBytes[5 - zeroCount] == Ip6::OptionPadN::kType))
+            {
+                padLength = zeroCount + 2;
```

Comment:
Okay i've changed the behavior to iterate through options.

I made the assumption before, that it doesn't metter if its part of non-pad option or not, since decompressor needs to add padding anyway. But I agree, that it should be done as you pointed.